src/wallet/scriptpubkeyman.cpp

```diff
@@ -1717,42 +1717,187 @@ std::unordered_set<CScript, SaltedSipHasher> LegacyDataSPKM::GetScriptPubKeys()
+        // If the redeemScript is not known, the P2SH output script is valid.
+        case TxoutType::SCRIPTHASH:
+        {
+            if (ctx != ScriptContext::TOP) return false;
+            CScriptID script_id = CScriptID(uint160(sols[0]));
+            CScript subscript;
+            if (GetCScript(script_id, subscript)) {
+                return is_valid_script(subscript, ScriptContext::P2SH);
+            }
+            return true;
+        }
+        // P2WSH is allowed as an output script or inside of P2SH, but not P2WSH.
+        // If the witnessScript is known, it must also be a valid script within P2WSH.
+        // If the witnessScript is not known, the P2WSH output script is valid.
+        case TxoutType::WITNESS_V0_SCRIPTHASH:
+        {
+            if (ctx == ScriptContext::P2WSH) return false;
+            // CScriptID is the hash160 of the script. For P2WSH, we already have the SHA256 of the script,
+            // so only RIPEMD160 of that is required to get the CScriptID for lookup.
+            CScriptID script_id{RIPEMD160(sols[0])};
+            CScript subscript;
+            if (GetCScript(script_id, subscript)) {
+                return is_valid_script(subscript, ScriptContext::P2WSH);
+            }
+            return true;
+        }
+        // Multisig in any nesting level is valid, but inside of P2WSH, all pubkeys must be compressed.
+        case TxoutType::MULTISIG:
+        {
+            if (ctx == ScriptContext::P2WSH) {
+                std::vector<std::vector<unsigned char>> keys(sols.begin() + 1, sols.begin() + sols.size() - 1);
+                if (!all_keys_compressed(keys)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        }
+        assert(false);
+    };
+    // Iterate again for all the P2WSH scripts
+    for (const auto& [id, script] : mapScripts) {
+        std::vector<std::vector<unsigned char>> sols;
+        TxoutType type = Solver(script, sols);
+        if (type == TxoutType::WITNESS_V0_SCRIPTHASH) {
+            CScript witness_script;
+            int wit_ver = -1;
+            std::vector<unsigned char> wit_prog;
+            // For a P2WSH output script to be spendable, we must know and inspect its witnessScript.
+            if (GetCScript(id, witness_script) &&
+                !witness_script.IsPayToScriptHash() && // P2SH inside of P2WSH is not allowed
+                !witness_script.IsWitnessProgram(wit_ver, wit_prog) && // Witness programs are not allowed inside of P2WSH
+                // We only allow scripts that we would consider spendable as an output script.
+                // Note that while this would exclude P2WSH-multisigs, we are already handling those in the first loop.
+                spks.count(witness_script) > 0 &&
+                // Pubkeys must be compressed.
+                is_valid_script(script, ScriptContext::P2WSH)) {
```

Comment:
I was able to contrive a test that requires this loop, which also revealed a bug.

The situation is if a `sh(wsh(pkh()))` is imported with `importmulti`, the `wsh(pkh())` will only be in `mapScripts`. However, this `wsh(pkh())` is still `ISMINE_SPENDABLE` and coins sent to it will be seen by the wallet. This loop is where that would be detected.